### ISSUE TITLE
manylinux ATLAS RPMs are much slower at SVD that debian / ubuntu's

### ISSUE THREAD

**Issue by ogrisel** [matthew-brett/manylinux-builds#4]
Here is a simple profiling output for numpy built with the manylinux-builds script (in the current manylinux docker environment):

```
>>> import numpy as np
%prun _ = np.linalg.svd(np.random.randn(500, 500))
         28 function calls in 3.748 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    3.733    3.733    3.733    3.733 linalg.py:1254(svd)
        1    0.015    0.015    0.015    0.015 {method 'randn' of 'mtrand.RandomState' objects}
        1    0.000    0.000    3.748    3.748 {built-in method builtins.exec}
        1    0.000    0.000    3.748    3.748 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 numeric.py:406(asarray)
        3    0.000    0.000    0.000    0.000 {method 'astype' of 'numpy.ndarray' objects}

In [4]: np
Out[4]: <module 'numpy' from '/usr/local/lib/python3.5/site-packages/numpy/__init__.py'>

In [5]: !ldd /usr/local/lib/python3.5/site-packages/numpy/core/multiarray.cpython-35m-x86_64-linux-gnu.so
    linux-vdso.so.1 (0x00007ffee5f00000)
    libtatlas.so.3 => /usr/local/lib/python3.5/site-packages/numpy/core/../.libs/libtatlas.so.3 (0x00007f2451af0000)
    libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f24517e7000)
    libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f24515c9000)
    libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f245121e000)
    /lib64/ld-linux-x86-64.so.2 (0x000055aec3f57000)
    libgfortran.so.3 => /usr/local/lib/python3.5/site-packages/numpy/core/../.libs/libgfortran.so.3 (0x00007f2450f2c000)
```

Here is the output using the system packages for `python3-numpy` and `libatlas3-base` on a recent ubuntu (in another docker container on the same host):

```
In [2]: %prun _ = np.linalg.svd(np.random.randn(500, 500))
         28 function calls in 0.322 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.294    0.294    0.294    0.294 linalg.py:1254(svd)
        1    0.028    0.028    0.028    0.028 {method 'randn' of 'mtrand.RandomState' objects}
        1    0.000    0.000    0.322    0.322 <string>:1(<module>)
        1    0.000    0.000    0.322    0.322 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 numeric.py:406(asarray)
        3    0.000    0.000    0.000    0.000 {method 'astype' of 'numpy.ndarray' objects}

In [3]: np
Out[3]: <module 'numpy' from '/usr/lib/python3/dist-packages/numpy/__init__.py'>

In [4]: !ldd /usr/lib/python3/dist-packages/numpy/core/multiarray.cpython-35m-x86_64-linux-gnu.so
    linux-vdso.so.1 =>  (0x00007ffebedf7000)
    libblas.so.3 => /usr/lib/libblas.so.3 (0x00007f07e20ab000)
    libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f07e1da2000)
    libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f07e1b84000)
    libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f07e17ba000)
    /lib64/ld-linux-x86-64.so.2 (0x000055788b782000)
    libatlas.so.3 => /usr/lib/libatlas.so.3 (0x00007f07e121c000)
    libgfortran.so.3 => /usr/lib/x86_64-linux-gnu/libgfortran.so.3 (0x00007f07e0eea000)
    libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f07e0cd4000)
    libquadmath.so.0 => /usr/lib/x86_64-linux-gnu/libquadmath.so.0 (0x00007f07e0a94000)
```

Here is the atlas package I used under ubuntu:

```
$ apt-cache show libatlas3-base
Package: libatlas3-base
Priority: optional
Section: universe/libs
Installed-Size: 10672
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Original-Maintainer: Debian Science Team <debian-science-maintainers@lists.alioth.debian.org>
Architecture: amd64
Source: atlas
Version: 3.10.2-9
Provides: libblas.so.3, liblapack.so.3
Depends: libc6 (>= 2.14), libgcc1 (>= 1:4.1.1), libgfortran3 (>= 4.6), libblas-common
Filename: pool/universe/a/atlas/libatlas3-base_3.10.2-9_amd64.deb
Size: 2696644
MD5sum: 48b0bac8a9211e79b0d2c739cc6f0f64
SHA1: 10309642d7d8718c9225e8101624f92bc268cf4e
SHA256: eddd960a129a4e32c21ec04b99ce7f6c172bfcd9e6e3d8c2f63f260764e4c479
Description: Automatically Tuned Linear Algebra Software, generic shared
Description-md5: f80833178f9c0623b9b635b55ecafa8d
Built-Using: lapack (= 3.5.0-5)
Homepage: http://math-atlas.sourceforge.net/
Bugs: https://bugs.launchpad.net/ubuntu/+filebug
Origin: Ubuntu
Supported: 9m
```

**Comment by ogrisel** [comment]
@matthew-brett I found the issue when running the scikit-learn tests for a scikit-learn package built using #2. They are much slower than usual because of slow SVD.

**Comment by matthew-brett** [comment]
Ouch - I was worried that might be true, but didn't think it would be that bad.

How about installing atlas via the centos repositories rather than my rebuild?

**Comment by ogrisel** [comment]
I have also tried to install `atlas` and `atlas-devel` (3.10) on centos 7 and it's very bad as well (in another docker container on the same host):

```
In [2]: import numpy as np

In [3]: %prun _ = np.linalg.svd(np.random.randn(500, 500))
         27 function calls in 4.341 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    4.321    4.321    4.321    4.321 linalg.py:1254(svd)
        1    0.020    0.020    0.020    0.020 {method 'randn' of 'mtrand.RandomState' objects}
        1    0.000    0.000    4.341    4.341 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 linalg.py:139(_commonType)
        3    0.000    0.000    0.000    0.000 {method 'astype' of 'numpy.ndarray' objects}

In [4]: np
Out[4]: <module 'numpy' from '/usr/lib64/python2.7/site-packages/numpy/__init__.pyc'>

In [5]: !ldd /usr/lib64/python2.7/site-packages/numpy/core/multiarray.so
    linux-vdso.so.1 =>  (0x00007ffc5d17a000)
    libtatlas.so.3 => /usr/lib64/atlas/libtatlas.so.3 (0x00007ff986fb7000)
    libm.so.6 => /lib64/libm.so.6 (0x00007ff986cb5000)
    libpython2.7.so.1.0 => /lib64/libpython2.7.so.1.0 (0x00007ff9868ee000)
    libpthread.so.0 => /lib64/libpthread.so.0 (0x00007ff9866d2000)
    libc.so.6 => /lib64/libc.so.6 (0x00007ff986311000)
    /lib64/ld-linux-x86-64.so.2 (0x00005562eb49b000)
```

**Comment by ogrisel** [comment]
I think we should have a look at the build options of the debian maintainers. They seem more knowledgeable about how to make a good binary for atlas :)

**Comment by ogrisel** [comment]
Actually I just tried to use the atlas 3.8 package from centos 5.11 (in the manylinux:x86_64 image) and the performance seems good to me:

```
In [1]: import numpy as np

In [2]: %prun _ = np.linalg.svd(np.random.randn(500, 500))
         28 function calls in 0.317 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.298    0.298    0.299    0.299 linalg.py:1254(svd)
        1    0.018    0.018    0.018    0.018 {method 'randn' of 'mtrand.RandomState' objects}
        1    0.000    0.000    0.317    0.317 <string>:1(<module>)
        1    0.000    0.000    0.317    0.317 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 numeric.py:406(asarray)
        3    0.000    0.000    0.000    0.000 {method 'astype' of 'numpy.ndarray' objects}

In [3]: np
Out[3]: <module 'numpy' from '/opt/_internal/cpython-3.5.1/lib/python3.5/site-packages/numpy/__init__.py'>

In [4]: !ldd /opt/_internal/cpython-3.5.1/lib/python3.5/site-packages/numpy/core/multiarray.cpython-35m-x86_64-linux-gnu.so
    linux-vdso.so.1 =>  (0x00007ffe9bab2000)
    libptf77blas.so.3 => /usr/lib64/atlas/libptf77blas.so.3 (0x00007f6b02ef4000)
    libptcblas.so.3 => /usr/lib64/atlas/libptcblas.so.3 (0x00007f6b02cd4000)
    libatlas.so.3 => /usr/lib64/atlas/libatlas.so.3 (0x00007f6b023df000)
    libm.so.6 => /lib64/libm.so.6 (0x00007f6b0215c000)
    libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f6b01f3f000)
    libc.so.6 => /lib64/libc.so.6 (0x00007f6b01be6000)
    /lib64/ld-linux-x86-64.so.2 (0x000055ca7d162000)
    libgfortran.so.1 => /usr/lib64/libgfortran.so.1 (0x00007f6b0194f000)
```

there is something weird in all the packages for atlas 3.10 on centos.

**Comment by matthew-brett** [comment]
You think it's reasonable to ship the wheels with the CentOS 5.11 atlas?

**Comment by ogrisel** [comment]
Summary:
- ubuntu / debian atlas 3.10 => GOOD
- centos 5.11 atlas 3.8 => GOOD
- your custom built centos 5.11 rpms fo atlas 3.10 => BAD
- official centos 7.2 atlas 3.10 => BAD

I would be in favor of just using centos 5.11 ATLAS packages. We need to rebuild numpy, scipy, pandas, scikit-learn and friends to check that their test suite pass and run at a reasonable speed before starting to upload anything to PyPI.

Unfortunately I will be AFK this WE.

**Comment by ogrisel** [comment]
> You think it's reasonable to ship the wheels with the CentOS 5.11 atlas?

Yes unless you feel the motivation and time to find a way to build a good atlas 3.10 on centos. I think this time would be more wisely invested in testing extensively an OpenBLAS experimental build of the full scipy stack.

**Comment by ogrisel** [comment]
It would be great if you could confirm that you observe the same performance behavior as reported in https://github.com/matthew-brett/manylinux-builds/issues/4#issuecomment-201359881 on one of your machines.

**Comment by matthew-brett** [comment]
I'm fine with not building another ATLAS :)

We should probably formulate some criteria for accepting an openblas build .

**Comment by ogrisel** [comment]
> We should probably formulate some criteria for accepting an openblas build . 

This should be discussed on the numpy / scipy MLs. But we could use your script to build an upload a full openblas-based stack to some external (non-pypi) http repo and tell people on the numpy ML to test them on their machines and report any test failure (or crash) in a comment thread somewhere to better inform the discussion.

**Comment by matthew-brett** [comment]
Hmm - no - I don't get the same as you - at least, installing the wheels outside the docker image.

Custom ATLAS wheels:

```
In [1]: import numpy as np

In [2]: %prun _ = np.linalg.svd(np.random.randn(500, 500))
         27 function calls in 0.309 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.281    0.281    0.281    0.281 linalg.py:1254(svd)
        1    0.028    0.028    0.028    0.028 {method 'randn' of 'mtrand.RandomState' objects}
        1    0.000    0.000    0.309    0.309 <string>:1(<module>)
```

Stock CentOS ATLAS wheels:

```
In [1]: import numpy as np

In [2]: %prun _ = np.linalg.svd(np.random.randn(500, 500))
         27 function calls in 0.346 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.318    0.318    0.318    0.318 linalg.py:1254(svd)
        1    0.028    0.028    0.028    0.028 {method 'randn' of 'mtrand.RandomState' objects}
        1    0.000    0.000    0.346    0.346 <string>:1(<module>)
```

Could your timing be an artefact of the docker container?   Do you get a difference outside the container?

**Comment by matthew-brett** [comment]
Here's the same for the machine on which the ATLAS RPM was built.

Custom ATLAS:

```
In [2]: %prun _ = np.linalg.svd(np.random.randn(500, 500))
         27 function calls in 0.481 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.461    0.461    0.461    0.461 linalg.py:1254(svd)
        1    0.020    0.020    0.020    0.020 {method 'randn' of 'mtrand.RandomState' objects}
        1    0.000    0.000    0.481    0.481 <string>:1(<module>)
```

CentOS ATLAS:

```
In [3]: %prun _ = np.linalg.svd(np.random.randn(500, 500))
         27 function calls in 0.348 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.328    0.328    0.328    0.328 linalg.py:1254(svd)
        1    0.020    0.020    0.020    0.020 {method 'randn' of 'mtrand.RandomState' objects}
        1    0.000    0.000    0.348    0.348 <string>:1(<module>)
```

All for Python 2.7.

**Comment by matthew-brett** [comment]
Similar results for Python 3.5.

**Comment by matthew-brett** [comment]
What do you get installing the custom ATLAS wheel into a Debian docker container?

I'm using : `pip install -f https://nipy.bic.berkeley.edu/manylinux numpy` (the manylinux wheels there are built with the custom ATLAS).

**Comment by matthew-brett** [comment]
I'm building both openblas and stock atlas numerical packages via travis-ci: https://travis-ci.org/matthew-brett/manylinux-builds/builds/118643111

The atlas builds get uploaded to http://f0a8b7f2b7d90646e9cd-f73b241aa77a5851fd7c16ebcb53c33c.r16.cf2.rackcdn.com and the openblas builds to http://ccdd0ebb5a931e58c7c5-aae005c4999d7244ac63632f8b80e089.r77.cf2.rackcdn.com

At a first pass, numpy, scipy, scikit-learn all passed their tests under openblas, pandas had quite a few errors, but they didn't look blas related.

**Comment by matthew-brett** [comment]
I guess, for OpenBLAS, it would be worth testing with the same processor list as the [OpenBLAS buildbots](https://github.com/xianyi/OpenBLAS/issues/785#issuecomment-193958360):
- core2
- nehalem
- sandybridge
- haswell
- barcelona
- bulldozer
- piledriver

on lines of `OPENBLAS_VERBOSE=2 OPENBLAS_CORETYPE=Core2 python
-c 'import scipy.linalg; scipy.linalg.test()'` (from https://mail.scipy.org/pipermail/numpy-discussion/2016-February/074889.html).

**Comment by matthew-brett** [comment]
I started some tests on the manylinux builds - see : https://github.com/matthew-brett/manylinux-testing

Tests on stock atlas, custom atlas break in various ways.   Tests on OpenBLAS appear to be rock solid on supported processors - see : https://travis-ci.org/matthew-brett/manylinux-testing/builds/118767771

OpenBLAS tests pass on processors up to (not including) haswell, piledriver, bulldozer, steamroller, which give illegal instructions.

I suspect these failures are just because the processor on travis is something less than a haswell, and we're forcing openblas to use the wrong kernel.

I added a job to test with automatic CPU detection here : https://travis-ci.org/matthew-brett/manylinux-testing/jobs/118780784 . It hasn't run yet, but I'm guessing it's going to pass.

So, I think OpenBLAS is the way to go, and we commit ourselves to working out any problems down the line, on the basis that these are going to be less than trying to support an ATLAS build.

**Comment by matthew-brett** [comment]
Statsmodels "full" test using MKL (https://travis-ci.org/statsmodels/statsmodels/jobs/118775648#L274) takes 482 seconds (https://travis-ci.org/statsmodels/statsmodels/jobs/118775648#L9123).  We can't run the full test (it times out on the slower kernels), but not "full" takes us 214 seconds (https://travis-ci.org/matthew-brett/manylinux-testing/jobs/118767778#L522) with the "Nehalem" kernel (https://travis-ci.org/matthew-brett/manylinux-testing/jobs/118767778#L343).  Obviously not directly comparable, but at least, not obviously out of ballpark.

**Comment by ogrisel** [comment]
> I suspect these failures are just because the processor on travis is something less than a haswell, and we're forcing openblas to use the wrong kernel.

Yes, you should not force a CPU architecture that you know is not the correct one. We should only test with automatic architecture detection on travis.

**Comment by matthew-brett** [comment]
Hmm - maybe I spoke too soon:

Statsmodels fails two tests with autodetection = sandybridge : 
- https://travis-ci.org/matthew-brett/manylinux-testing/jobs/119188668#L7819
- https://travis-ci.org/matthew-brett/manylinux-testing/jobs/119188675#L5107

```
======================================================================
FAIL: statsmodels.tsa.statespace.tests.test_dynamic_factor.TestDynamicFactor_general_errors.test_bse_approx
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/matthew-brett/manylinux-testing/venv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/matthew-brett/manylinux-testing/venv/lib/python2.7/site-packages/statsmodels/tsa/statespace/tests/test_dynamic_factor.py", line 376, in test_bse_approx
    assert_allclose(bse[-10:], self.true['var_oim'][-10:], atol=1e-4)
  File "/home/travis/build/matthew-brett/manylinux-testing/venv/lib/python2.7/site-packages/numpy/testing/utils.py", line 1391, in assert_allclose
    verbose=verbose, header=header)
  File "/home/travis/build/matthew-brett/manylinux-testing/venv/lib/python2.7/site-packages/numpy/testing/utils.py", line 733, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0.0001
(mismatch 10.0%)
 x: array([  2.142085e-04,   1.307592e-02,   2.915582e-01,   4.320433e-01,
         7.689779e-04,   1.742122e-02,   2.201651e-02,   5.543138e-04,
         1.456134e-02,   1.707097e-02])
 y: array([  2.142100e-04,   1.307587e-02,   2.916752e-01,   4.320406e-01,
         7.689900e-04,   1.742173e-02,   2.201610e-02,   5.543500e-04,
         1.456365e-02,   1.707167e-02])
======================================================================
FAIL: statsmodels.tsa.statespace.tests.test_sarimax.TestFriedmanStateRegression.test_mle
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/matthew-brett/manylinux-testing/venv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/matthew-brett/manylinux-testing/venv/lib/python2.7/site-packages/statsmodels/tsa/statespace/tests/test_sarimax.py", line 746, in test_mle
    atol=1e-1, rtol=1e-1
  File "/home/travis/build/matthew-brett/manylinux-testing/venv/lib/python2.7/site-packages/numpy/testing/utils.py", line 1391, in assert_allclose
    verbose=verbose, header=header)
  File "/home/travis/build/matthew-brett/manylinux-testing/venv/lib/python2.7/site-packages/numpy/testing/utils.py", line 733, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=0.1, atol=0.1
(mismatch 33.3333333333%)
 x: array([   0.970836,    0.299662,  103.355682])
 y: array([  0.934849,   0.309059,  93.224973])
----------------------------------------------------------------------
Ran 4548 tests in 2554.303s
FAILED (SKIP=164, failures=2)
```

Passes with:
- prescott : https://travis-ci.org/matthew-brett/manylinux-testing/jobs/119188670#L7820
- core2 : https://travis-ci.org/matthew-brett/manylinux-testing/jobs/119188673#L5107
- dunnington : https://travis-ci.org/matthew-brett/manylinux-testing/jobs/119188671#L7822
- penryn : https://travis-ci.org/matthew-brett/manylinux-testing/jobs/119188672#L7820
- nehalem : https://travis-ci.org/matthew-brett/manylinux-testing/jobs/119188674#L7822

**Comment by ogrisel** [comment]
Maybe we could ask @xianyi if he would be interested in running the numpy, scikit-learn and statsmodel tests on the https://github.com/xianyi/OpenBLAS-CI infrastructure in addition to scipy? This build farm has access to different generations of CPUs not available on travis.

**Comment by matthew-brett** [comment]
I just set up more or less the same tests on a couple of machines I have - so that should be pretty easy:

http://nipy.bic.berkeley.edu/builders/manylinux-2.7-debian/builds/7

http://nipy.bic.berkeley.edu/builders/manylinux-2.7-fedora/builds/2

**Comment by ogrisel** [comment]
@matthew-brett the statsmodel test failures you found with openblas auto-detection to sandybridge look similar to issue statsmodel/statsmodel#2837 reported by @josef-pkt. They might reveal a numerical stability issue in one of the OpenBLAS kernels for the sandybridge architecture.

**Comment by xianyi** [comment]
@ogrisel , @matthew-brett , I like to run other tests.

**Comment by matthew-brett** [comment]
@xianyi - that would be very helpful - to run the numerical Python tests on the buildbots.

How can I help?  Is your 'master.cfg' somewhere that I can edit, to set that up?   I would be happy to make a PR for you to look at.

**Comment by ogrisel** [comment]
numpy, scipy and scikit-learn tests are now run on  https://github.com/xianyi/OpenBLAS-CI . However the statsmodels tests are still not run there. @matthew-brett can you still reproduce the above statsmodels tests failures with the latest version of OpenBLAS? If so we should probably also make the statsmodels tests part of the OpenBLAS-CI "full" suite.

**Comment by matthew-brett** [comment]
Sorry - I should have got back to you.  In the discussion over at https://github.com/statsmodels/statsmodels/issues/2876 the author of the relevant code put this down to acceptable precision differences, and it seems that tracking down the exact difference would be very hard, so we relaxed the test threshold - see https://github.com/statsmodels/statsmodels/pull/2878.  The test passes now on my buildbots : http://nipy.bic.berkeley.edu/builders/manylinux-2.7-fedora/builds/40/steps/shell_14/logs/stdio

**Comment by xianyi** [comment]
@ogrisel , @matthew-brett , I plan to release a new version this week.

**Comment by ogrisel** [comment]
@matthew-brett  ok great. I think we can close this issue as "wontfix" then. OpenBLAS seems stable enough to fully support the scipy stack under linux.

**Comment by matthew-brett** [comment]
Finally closing with won't fix :)
